src/dmd/aliasthis.d

```diff
@@ -31,17 +32,12 @@ import dmd.visitor;
  */
 extern (C++) final class AliasThis : Dsymbol
 {
-    Identifier ident;
     /// The symbol this `alias this` resolves to
     Dsymbol sym;
     /// Whether this `alias this` is deprecated or not
     bool isDeprecated_;
 
-    extern (D) this(const ref Loc loc, Identifier ident)
-    {
-        super(loc, null);    // it's anonymous (no identifier)
-        this.ident = ident;
-    }
+    mixin parseTimePropertiesAliasThis;
```

Comment:
This doesn't really address the duplication though, does it?  There are still two copies of `AliasThis` present, and every time a new AST node is introduced it must still be defined twice.

Reply:
The only way we could eliminate the duplication would be to have the ASTCodegen classes inherit from the ASTBase classes, however, that would require multiple inheritance since a node must inherit its parse time counterpart but also its logical ancestor (for example AddExp would need to inherit ParseTimeAddExp and BinaryExp). This could technically be done by using inheritance and alias this, but I would rather not go there since this currently isn't defined behavior (and ideally alias this would be deprecated for classes).

Instead, this PR limits the scope of the duplication. Indeed, new AST node declarations still need to be duplicated, but I would argue that this doesn't happen very often.

I couldn't come up with a solution that doesn't require to refactor the compiler entirely (e.g. stripping all AST nodes of their functions, basically transforming them into data containers) . If anyone has any propositions, I'm open to suggestions. However, I doubt that a feasible solution to get rid of the duplicated AST node declarations is possible at this point.

Reply:
> I couldn't come up with a solution that doesn't require to refactor the compiler entirely (e.g. stripping all AST nodes of their functions, basically transforming them into data containers) . If anyone has any propositions, I'm open to suggestions. However, I doubt that a feasible solution to get rid of the duplicated AST node declarations is possible at this point.

Why not? We've already removed the semantic, semantic2, semantic3, resolve, toIR, toCtype, toElem, toDt, and probably many more this way.

1. There are pointless member functions:
https://github.com/dlang/dmd/blob/0ff19fbade6faceeb8b94bf53ee616c1cabf2535/src/dmd/statement.d#L216-L249

2. There doesn't need to be a big clear-out done, just anything that isn't to do with examining itself. e.g. from Statement:
https://github.com/dlang/dmd/blob/0ff19fbade6faceeb8b94bf53ee616c1cabf2535/src/dmd/statement.h#L119-L129
Remove the above highlighted functions, and astbase can just publicly import `dmd.statement`...

Reply:
Unfortunately, it's not that easy. Looking at `statement.d` it imports a lot of modules that don't have anything to do with parsing. Before being able to import `statement.d` we need to get rid of all of those dependencies. That is no easy task and it requires a lot of refactorings.

Reply:
I'd be surprised if it stays that way after dealing with the above functions.

Reply:
The problem is that `statement.d` has many dependencies. Even if we move out everything that relies on semantic information from statement.d, we will still have some Expression fields. `expression.d` has other dependencies that rely on semantic information, so now we need to the same for it too etc. My point is that, before we can simply do `import statement;` in ASTBase, we need to do this cleanup for all the AST nodes, otherwise the import dependencies is going to drag semantic information into ASTBase.